gemini-controller panics on K8s v1.22.6

### What happened?

All worked just fine on various K8s v1.21.x versions, however after update to K8s v1.22.6 (GKE version v1.22.6-gke.300), gemini panics on start

```
I0215 09:33:27.399825       1 main.go:18] Running in verbose mode
I0215 09:33:27.899155       1 request.go:530] Throttling request took 99.114245ms, request: GET:https://192.168.196.1:443/apis/batch/v1?timeout=32s
I0215 09:33:28.099217       1 request.go:530] Throttling request took 299.09222ms, request: GET:https://192.168.196.1:443/apis/policy/v1?timeout=32s
I0215 09:33:28.099677       1 request.go:530] Throttling request took 299.638954ms, request: GET:https://192.168.196.1:443/apis/batch/v1beta1?timeout=32s
I0215 09:33:28.199158       1 request.go:530] Throttling request took 399.099267ms, request: GET:https://192.168.196.1:443/apis/certificates.k8s.io/v1?timeout=32s
I0215 09:33:28.199421       1 request.go:530] Throttling request took 399.357316ms, request: GET:https://192.168.196.1:443/apis/networking.k8s.io/v1?timeout=32s
I0215 09:33:28.199689       1 request.go:530] Throttling request took 399.480971ms, request: GET:https://192.168.196.1:443/apis/storage.k8s.io/v1beta1?timeout=32s
I0215 09:33:28.200129       1 request.go:530] Throttling request took 399.808225ms, request: GET:https://192.168.196.1:443/apis/scheduling.k8s.io/v1?timeout=32s
I0215 09:33:28.349008       1 request.go:530] Throttling request took 249.806937ms, request: GET:https://192.168.196.1:443/apis/policy/v1beta1?timeout=32s
I0215 09:33:28.350454       1 request.go:530] Throttling request took 250.770056ms, request: GET:https://192.168.196.1:443/apis/rbac.authorization.k8s.io/v1?timeout=32s
I0215 09:33:28.399210       1 request.go:530] Throttling request took 498.899081ms, request: GET:https://192.168.196.1:443/apis/admissionregistration.k8s.io/v1?timeout=32s
I0215 09:33:28.399677       1 request.go:530] Throttling request took 549.117845ms, request: GET:https://192.168.196.1:443/apis/apiextensions.k8s.io/v1?timeout=32s
panic: the server could not find the requested resource

goroutine 1 [running]:
github.com/fairwindsops/gemini/pkg/kube.createClient(0x42)
	/go/src/github.com/fairwindsops/gemini/pkg/kube/client.go:92 +0x66d
github.com/fairwindsops/gemini/pkg/kube.GetClient(...)
	/go/src/github.com/fairwindsops/gemini/pkg/kube/client.go:59
github.com/fairwindsops/gemini/pkg/controller.NewController(0x1d88c20)
	/go/src/github.com/fairwindsops/gemini/pkg/controller/controller.go:72 +0x5e6
main.main()
	/go/src/github.com/fairwindsops/gemini/main.go:19 +0x3d
```

### What did you expect to happen?

Most likely this is related to the `beta CustomResourceDefinition` a.k.a. `apiextensions.k8s.io/v1beta1` [API removal](https://kubernetes.io/blog/2021/07/14/upcoming-changes-in-kubernetes-1-22/#api-changes).

### How can we reproduce this?

Deploy to K8s v1.22.x version.

### version

0.1

### Search

- [X] I did search for other open and closed issues before opening this.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

As a fix, I can change https://github.com/FairwindsOps/gemini/blob/3e3cc90c8a38af59ed8f179201883e93d3021304/pkg/kube/client.go#L92 to use `extClientSet.ApiextensionsV1().CustomResourceDefinitions()` and change https://github.com/FairwindsOps/gemini/blob/3e3cc90c8a38af59ed8f179201883e93d3021304/pkg/kube/client.go#L118 to use first served version. This most likely would not be a breaking change as `apiextensions.k8s.io/v1` API is available since v1.16.

--- a/pkg/kube/client.go
+++ b/pkg/kube/client.go
@@ -16,8 +16,10 @@ package kube
 
 import (
 	"context"
+	"errors"
 	"time"
 
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
@@ -89,10 +91,15 @@ func createClient() *Client {
 		panic(err)
 	}
 	restMapper := restmapper.NewDiscoveryRESTMapper(resources)
-	snapshotCRD, err := extClientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), "volumesnapshots."+VolumeSnapshotGroupName, metav1.GetOptions{})
+	snapshotCRD, err := extClientSet.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), "volumesnapshots."+VolumeSnapshotGroupName, metav1.GetOptions{})
 	if err != nil {
 		panic(err)
 	}
+	volumeSnapshotVersion, err := getVolumeSnapshotVersion(snapshotCRD.Spec.Versions)
+	if err != nil {
+		panic(err)
+	}
+
 	vsMapping, err := restMapper.RESTMapping(schema.GroupKind{
 		Group: VolumeSnapshotGroupName,
 		Kind:  VolumeSnapshotKind,
@@ -115,6 +122,15 @@ func createClient() *Client {
 		InformerFactory:       informerFactory,
 		SnapshotClient:        snapshotClient,
 		SnapshotGroupClient:   sgClientSet.SnapshotgroupV1beta1(),
-		VolumeSnapshotVersion: VolumeSnapshotGroupName + "/" + snapshotCRD.Spec.Version,
+		VolumeSnapshotVersion: VolumeSnapshotGroupName + "/" + volumeSnapshotVersion,
 	}
+}
+
+func getVolumeSnapshotVersion(v []v1.CustomResourceDefinitionVersion) (string, error) {
+	for _, crd := range v {
+		if crd.Served {
+			return crd.Name, nil
+		}
+	}
+	return "", errors.New("no " + VolumeSnapshotGroupName + " served API found")
 }

--- a/pkg/kube/fake_client.go
+++ b/pkg/kube/fake_client.go
@@ -36,7 +36,7 @@ func SetFakeClient() *Client {
 }
 
 func createFakeClient() *Client {
-	objects := []k8sruntime.Object{}
+	var objects []k8sruntime.Object
 	k8s := k8sfake.NewSimpleClientset(objects...)
 	_ = snapshotsFake.NewSimpleClientset(objects...)
 

--- a/pkg/types/snapshotgroup/v1beta1/crd.go
+++ b/pkg/types/snapshotgroup/v1beta1/crd.go
@@ -6,7 +6,7 @@ import (
 	"reflect"
 	"time"
 
-	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -17,65 +17,57 @@ import (
 
 // CreateCustomResourceDefinition creates the CRD and add it into Kubernetes. If there is error,
 // it will do some clean up.
-func CreateCustomResourceDefinition(namespace string, clientSet apiextensionsclientset.Interface) (*apiextensionsv1beta1.CustomResourceDefinition, error) {
-	crd := &apiextensionsv1beta1.CustomResourceDefinition{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      CRDName,
-			Namespace: namespace,
-		},
-		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
-			Group:   GroupName,
-			Version: SchemeGroupVersion.Version,
-			Scope:   apiextensionsv1beta1.NamespaceScoped,
-			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
-				Plural: Plural,
-				Kind:   reflect.TypeOf(SnapshotGroup{}).Name(),
-			},
-			Validation: &apiextensionsv1beta1.CustomResourceValidation{
-				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
-					Type: "object",
-					Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-						"spec": {
-							Type: "object",
-							Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-								"persistentVolumeClaim": {
-									Type: "object",
-									Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-										"claimName": {
-											Description: "PersistentVolumeClaim name to backup",
-											Type:        "string",
-										},
-										"spec": {
-											Description: "PersistentVolumeClaim spec to create and backup",
-											Type:        "object",
-										},
+func CreateCustomResourceDefinition(namespace string, clientSet apiextensionsclientset.Interface) (*apiextensionsv1.CustomResourceDefinition, error) {
+	crdVersion := apiextensionsv1.CustomResourceDefinitionVersion{
+		Name:               GroupVersion,
+		Served:             true,
+		Storage:            true,
+		Deprecated:         false,
+		DeprecationWarning: nil,
+		Schema: &apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type: "object",
+				Properties: map[string]apiextensionsv1.JSONSchemaProps{
+					"spec": {
+						Type: "object",
+						Properties: map[string]apiextensionsv1.JSONSchemaProps{
+							"persistentVolumeClaim": {
+								Type: "object",
+								Properties: map[string]apiextensionsv1.JSONSchemaProps{
+									"claimName": {
+										Description: "PersistentVolumeClaim name to backup",
+										Type:        "string",
+									},
+									"spec": {
+										Description: "PersistentVolumeClaim spec to create and backup",
+										Type:        "object",
 									},
 								},
-								"schedule": {
-									Type: "array",
-									Items: &apiextensionsv1beta1.JSONSchemaPropsOrArray{
-										Schema: &apiextensionsv1beta1.JSONSchemaProps{
-											Type: "object",
-											Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-												"every": {
-													Description: "Interval for creating new backups",
-													Type:        "string",
-												},
-												"keep": {
-													Description: "Number of historical backups to keep",
-													Type:        "integer",
-												},
+							},
+							"schedule": {
+								Type: "array",
+								Items: &apiextensionsv1.JSONSchemaPropsOrArray{
+									Schema: &apiextensionsv1.JSONSchemaProps{
+										Type: "object",
+										Properties: map[string]apiextensionsv1.JSONSchemaProps{
+											"every": {
+												Description: "Interval for creating new backups",
+												Type:        "string",
+											},
+											"keep": {
+												Description: "Number of historical backups to keep",
+												Type:        "integer",
 											},
 										},
 									},
 								},
-								"template": {
-									Type: "object",
-									Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-										"spec": {
-											Description: "VolumeSnapshot spec",
-											Type:        "object",
-										},
+							},
+							"template": {
+								Type: "object",
+								Properties: map[string]apiextensionsv1.JSONSchemaProps{
+									"spec": {
+										Description: "VolumeSnapshot spec",
+										Type:        "object",
 									},
 								},
 							},
@@ -84,8 +76,28 @@ func CreateCustomResourceDefinition(namespace string, clientSet apiextensionscli
 				},
 			},
 		},
+		Subresources:             nil,
+		AdditionalPrinterColumns: nil,
 	}
-	_, err := clientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{})
+
+	crd := &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      CRDName,
+			Namespace: namespace,
+		},
+		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+			Group: GroupName,
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Plural: Plural,
+				Kind:   reflect.TypeOf(SnapshotGroup{}).Name(),
+			},
+			Scope:                 apiextensionsv1.NamespaceScoped,
+			Versions:              []apiextensionsv1.CustomResourceDefinitionVersion{crdVersion},
+			PreserveUnknownFields: false,
+		},
+	}
+
+	_, err := clientSet.ApiextensionsV1().CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{})
 	if err == nil {
 		fmt.Println("CRD SnapshotGroup is created")
 	} else if apierrors.IsAlreadyExists(err) {
@@ -98,7 +110,7 @@ func CreateCustomResourceDefinition(namespace string, clientSet apiextensionscli
 
 	// Wait for CRD creation.
 	err = wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
-		crd, err = clientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), CRDName, metav1.GetOptions{})
+		crd, err = clientSet.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), CRDName, metav1.GetOptions{})
 		if err != nil {
 			fmt.Printf("Fail to wait for CRD SnapshotGroup creation: %+v\n", err)
 
@@ -106,12 +118,12 @@ func CreateCustomResourceDefinition(namespace string, clientSet apiextensionscli
 		}
 		for _, cond := range crd.Status.Conditions {
 			switch cond.Type {
-			case apiextensionsv1beta1.Established:
-				if cond.Status == apiextensionsv1beta1.ConditionTrue {
+			case apiextensionsv1.Established:
+				if cond.Status == apiextensionsv1.ConditionTrue {
 					return true, err
 				}
-			case apiextensionsv1beta1.NamesAccepted:
-				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+			case apiextensionsv1.NamesAccepted:
+				if cond.Status == apiextensionsv1.ConditionFalse {
 					fmt.Printf("Name conflict while wait for CRD SnapshotGroup creation: %s, %+v\n", cond.Reason, err)
 				}
 			}
@@ -123,7 +135,7 @@ func CreateCustomResourceDefinition(namespace string, clientSet apiextensionscli
 	// If there is an error, delete the object to keep it clean.
 	if err != nil {
 		fmt.Println("Try to cleanup")
-		deleteErr := clientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(context.TODO(), CRDName, metav1.DeleteOptions{})
+		deleteErr := clientSet.ApiextensionsV1().CustomResourceDefinitions().Delete(context.TODO(), CRDName, metav1.DeleteOptions{})
 		if deleteErr != nil {
 			fmt.Printf("Fail to delete CRD SnapshotGroup: %+v\n", deleteErr)
 

